kolibri/core/deviceadmin/utils.py

```diff
@@ -179,3 +181,37 @@ def search_latest(search_root, fallback_version):
 
     if newest:
         return os.path.join(search_root, newest)
+
+
+def perform_vacuum(database=db.DEFAULT_DB_ALIAS):
+    connection = db.connections[database]
+    if connection.vendor == "sqlite":
+        try:
+            db.close_old_connections()
+            db.connections.close_all()
+            cursor = connection.cursor()
+            cursor.execute("vacuum;")
+            connection.close()
+        except Exception as e:
+            logger.error(e)
+            new_msg = (
+                "Vacuum of database {db_name} couldn't be executed. Possible reasons:\n"
+                "  * There is an open transaction in the db.\n"
+                "  * There are one or more active SQL statements.\n"
+                "The full error: {error_msg}"
+            ).format(
+                db_name=db.connections[database].settings_dict["NAME"], error_msg=e
+            )
+            logger.error(new_msg)
+        else:
+            logger.info("Sqlite database Vacuum finished.")
+
+
+def schedule_vacuum():
+    current_dt = local_now()
+    vacuum_time = current_dt.replace(hour=3, minute=0, second=0, microsecond=0)
+    if vacuum_time < current_dt:
+        # If it is past 3AM, change the day to tomorrow.
+        vacuum_time = vacuum_time.replace(day=vacuum_time.day + 1)
```

Comment:
Got an error when the date today is the last day of the month. (i.e. Oct 31st)
Here is part of the error traceback:
```
File "kolibri/utils/server.py", line 90, in run_services
    schedule_vacuum()
  File "kolibri/core/deviceadmin/utils.py", line 217, in schedule_vacuum
    vacuum_time = vacuum_time.replace(day=vacuum_time.day + 1)
ValueError: day is out of range for month
```

Reply:
Ah, always good to test on an edge case day, nice catch!

Reply:
This should be working correctly now, using a timedelta.